CRM/Financial/BAO/FinancialType.php

```diff
@@ -181,11 +181,102 @@ public static function getIncomeFinancialType() {
     );
 
     foreach ($financialType as $key => $financialTypeName) {
-      if (!in_array($key, $revenueFinancialType)) {
+      if (!in_array($key, $revenueFinancialType) || !CRM_Core_Permission::check('add contributions of type ' . $financialTypeName)) {
         unset($financialType[$key]);
       }
+      
     }
     return $financialType;
   }
 
+  /**
+   * adding permissions for financial types
+   *
+   *
+   * @param array $permissions        
+   *   an array of permissions
+   */
+  public static function permissionedFinancialTypes(&$permissions) {
+    $financialTypes = CRM_Contribute_PseudoConstant::financialType();
+    $prefix = ts('CiviCRM') . ': ';
+    foreach ($financialTypes as $id => $type) {
+      $permissions['add contributions of type ' . $type] = $prefix . ts('add contributions of type ' . $type);
+      $permissions['view contributions of type ' . $type] = $prefix . ts('view contributions of type ' . $type);
+      $permissions['edit contributions of type ' . $type] = $prefix . ts('edit contributions of type ' . $type);
+      $permissions['delete contributions of type ' . $type] = $prefix . ts('delete contributions of type ' . $type);
+    }
+    $permissions['administer CiviCRM Financial Types'] = $prefix . ts('administer CiviCRM Financial Types');
+  }
+
+  public static function getAvailableFinancialTypes(&$financialTypes = NULL, $action = 'view') {
```

Comment:
Joe, $action is indeed referenced in the function. By default, it is 'view' but in certain instances, we can change it to check for various CRUD permissions by changing $action while invoking.